File: 2-Regression/2-Multiple Linear Regression/multiple_linear_regresiion.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sat Mar 31 11:39:45 2018

@author: kifli
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.cross_validation import train_test_split

from sklearn.linear_model import LinearRegression


def backwardElimination(X,SL):
    pvalue = 1
    x_column = np.arange(len(X[0]))
    
    while(pvalue > SL):
        X_opt = X[:,x_column]
        regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
        pvalues = regressor_OLS.pvalues        
        indice = np.where(pvalues == pvalues.max())
        pvalue = pvalues[indice]
        if (pvalue > SL):
            x_column = np.delete(x_column,indice)
            
    return x_column
    #print(regressor_OLS.summary())

def backwardELiminationwithAdjR(X,SL):
    var = True
    x_column = np.arange(len(X[0]))
    
    while var:
        X_opt = X[:,x_column]
        regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
        pvalues = regressor_OLS.pvalues        
        adjR_before = regressor_OLS.rsquared_adj.astype(float)
        indice = np.where(pvalues == pvalues.max())
        pvalue = pvalues[indice]
        if (pvalue > SL):
            x_column = np.delete(x_column,indice,1)
            X_opt = X[:,x_column]
            temp_regressor = sm.OLS(y,X_opt).fit()
            adjR_after = temp_regressor.rsquared_adj.astype(float)
            if adjR_before >= adjR_after:
                x_rollback = np.hstack((x, temp[:,[0,j]]))
                x_rollback = np.delete(x_rollback, j, 1)
                        
        else:
            var = False

def backwardEliminationWithR2(x, SL):
    numVars = len(x[0])
    temp = np.zeros((50,6)).astype(int)
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues).astype(float)
        adjR_before = regressor_OLS.rsquared_adj.astype(float)
        if maxVar > SL:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    temp[:,j] = x[:, j]
                    x = np.delete(x, j, 1)
                    tmp_regressor = sm.OLS(y, x).fit()
                    adjR_after = tmp_regressor.rsquared_adj.astype(float)
                    if (adjR_before >= adjR_after):
                        x_rollback = np.hstack((x, temp[:,[0,j]]))
                        x_rollback = np.delete(x_rollback, j, 1)
                        print (regressor_OLS.summary())
                        return x_rollback
                    else:
                        continue
    regressor_OLS.summary()
    return x

dataset = pd.read_csv('50_Startups.csv')

# Setting independent and dependent variables into X and y
X = dataset.iloc[:,:-1].values
y = dataset.iloc[:,-1].values

# Categorical variables
labelEncoder_x = LabelEncoder()
X[:,3] = labelEncoder_x.fit_transform(X[:,3])

onehotencoder = OneHotEncoder(categorical_features=[3])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the dummy variable trap
X = X[:,1:]

# Splitting the data set into the Training set and Test set
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state = 0)

regressor = LinearRegression()
regressor.fit(X_train,y_train)

# Predicting the Test set results
y_pred = regressor.predict(X_test)

# Building the optimal model using Backward Elimination
import statsmodels.formula.api as sm
X = np.append(np.ones((len(X[:,0]),1)).astype(int),X, 1)
SL = 0.05
backwardElimination(X,SL)

# Select Significance Level (SL) : 0.05
#regressor_OLS.summary()

File: 2-Regression/2-Multiple Linear Regression/multiple_linear_regresiion.R
# Multiple Linear Regression

# add dataset into R workspace
dataset = read.csv('50_Startups.csv')

# Encoding categorical the dataset
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1,2,3))

# Splitting dataset into the Training set and Test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

regressor = lm(formula = Profit ~ .,
               data = training_set)

y_pred1 = predict(regressor, newdata = test_set)

regressor = lm(formula = Profit ~ R.ACY.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)
summary(regressor)


regressor = lm(formula = Profit ~ R.ACY.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)


regressor = lm(formula = Profit ~ R.ACY.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)


regressor = lm(formula = Profit ~ R.ACY.D.Spend,
               data = dataset)
summary(regressor)
# since the most significance value only R.D.Spent then, equation
# can be changed into

regressor = lm(formula = Profit ~ R.ACY.D.Spend,
               data = training_set)

y_pred2 = predict(regressor, newdata = test_set)

backwardElimination <- function(x, sl) {
  numVars = length(x)
  for (i in c(1:numVars)){
    regressor = lm(formula = Profit ~ ., data = x)
    maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
    if (maxVar > sl){
      j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
      x = x[, -j]
    }
    numVars = numVars - 1
  }
  return(summary(regressor))
}

SL = 0.05
dataset = dataset[, c(1,2,3,4,5)]
backwardElimination(training_set, SL)
